sycl/doc/extensions/proposed/sycl_ext_oneapi_launch_queries.asciidoc

```diff
@@ -0,0 +1,263 @@
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+This extension introduces a new mechanism for querying any limitations that a
+developer must respect when launching a specific kernel on a specific queue.
+Such limitations may exist when a kernel is decorated with one or more
+properties that require an implementation to enable additional features
+(such as providing certain forward progress guarantees, or enabling
+cross-work-group synchronization routines), or when a kernel uses certain
+features (such as static work-group local memory or group algorithms).
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_LAUNCH_QUERIES` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== Launch queries
+
+An implementation's ability to satisify requests for specific behaviors
+(such as strong forward progress guarantees and/or cross-work-group
+synchronization) for a kernel may be dependent upon: the kernel itself, and the
+features that it uses; the queue (and underlying device) to which the kernel is
+submitted; and the kernel's launch configuration.
+
+It is a user's responsibility to ensure that a kernel requesting specific
+behaviors uses a compatible launch configuration, using the
+`ext_oneapi_get_info` function and descriptors from the `info::kernel`
+namespace.
+
+If the `sycl::nd_range` parameter used to launch a kernel is incompatible with
+the results of a kernel's launch queries, an implementation must throw a
+synchronous exception with the `errc::nd_range` error code.
+
+[NOTE]
+====
+The values returned by `ext_oneapi_get_info` account for all properties
+attached to a kernel (via the mechanisms defined in the
+sycl_ext_oneapi_kernel_properties extension), as well as the usage of features
+like group algorithms and work-group local memory. Developers should assume
+that the values will differ across kernels.
+====
+
+[source,c++]
+----
+namespace sycl {
+
+class kernel {
+ public:
+  template <typename Param, typename... T>
+  typename Param::return_type ext_oneapi_get_info(T... args) const;
+};
+
+}
+----
+
+[source,c++]
+----
+template <typename Param, typename... T>
+typename Param::return_type ext_oneapi_get_info(T... args) const;
+----
+_Constraints_: Available only when the types `+T...+` described by the parameter
+pack match the type list defined by `Param::argument_types`.
```

Comment:
It's a little unclear whether `Param::argument_types` is part of the specification or if it is "exposition only".  There is a reference to this type alias here, but it's not specified anywhere.  Which did you intend?

Reply:
@rolandschulz convinced me that we don't want to specify exactly how the type list is implemented, only that it is a type list.  I've tried to make this clearer in f9570bb8a402, by saying that the argument types in the list must match the table, and that the actual representation of the list is unspecified.